src/java.base/share/classes/java/security/Provider.java

```diff
@@ -1267,18 +1319,14 @@ public Service getService(String type, String algorithm) {
      *
      * @since 1.5
      */
-    public synchronized Set<Service> getServices() {
+    public Set<Service> getServices() {
         checkInitialized();
-        if (legacyChanged || servicesChanged) {
-            serviceSet = null;
-        }
-        if (serviceSet == null) {
-            ensureLegacyParsed();
+        if (serviceSet == null || legacyChanged || servicesChanged) {
```

Comment:
`serviceSet` should be at least made `volatile`

Reply:
Sure, will fix.